### ISSUE TITLE
bring back slick 2's blocking api

### ISSUE THREAD

**Issue by cvogt** [slick/slick#1564]
by porting it to master. No plans to implement this by the core maintainers, but we'd be supportive of a community effort.

This would be in addition to the reactive api of course.

**Comment by nafg** [comment]
What changed?

**Comment by cvogt** [comment]
@nafg there was more demand for the blocking api as an alternative than expected.

**Comment by takezoe** [comment]
I'm making a utility that add blocking execution methods to `Query` by implicit conversion.
https://gist.github.com/takezoe/28f7a462666d8969fd7433798774fc2f

What do you think about this?

**Comment by cvogt** [comment]
interesting @takezoe ! how would the usage look like?

**Comment by takezoe** [comment]
@cvogt Thanks for your interest. Usage is below:
```scala
import BlockingH2Driver._
import BlockingH2Driver.api._

implicit val session = ...

val users: Seq[UserRow] = Users.sortBy(_.id).list
val userOption: Option[UserRow] = Users.filter(_.id === id.bind).firstOption
val user: UserRow = Users.filter(_.id === id.bind).first

val insertedCount = Users.unsafeInsert(UserRow(1, "takezoe"))
val updatedCount = Users.filter(_id === id.bind).map(_.name).unsafeUpdate("takezoe")
val deletedCount = Users.filter(_.id === id.bind).unsafeDelete
```
We can execute `Query` as same as Slick2 using an implicit conversion which is provided by `SlickBlockingAPI`, but `+=`, `update` and `delete` method are conflict with action creation methods, so I gave other name like `unsafeXxxx` to the blocking execution methods.

**Comment by cvogt** [comment]
@takezoe where does the session come from?

**Comment by takezoe** [comment]
@cvogt I'm using session created by `db.createSession()`.

**Comment by takezoe** [comment]
@cvogt Here is a complete example of blocking API with Play2:
https://github.com/takezoe/blocking-slick-play2

implicit conversions are defined in:
https://github.com/takezoe/blocking-slick-play2/blob/master/app/slick/SlickBlockingAPI.scala

usage in the controller:
https://github.com/takezoe/blocking-slick-play2/blob/master/app/controllers/UserController.scala

Of course, this blocking API is prototype which is made for moving our Slick2 based applications to Slick3, so I would like to hear your opinion.

**Comment by cvogt** [comment]
Very interesting @takezoe. Are you interested in contributing this back into Slick? Maybe as a separate project depending on slick core to not make slick core even bigger than it is? But I think we could totally host this under the slick github space, e.g. `github.com/slick/slick-blocking` and make you the admin. And we can also make any changes necessary to Slick itself in order to allow the api to be as source compatible to slick 2 as possible - e.g. allow bringing back "withSession". WDYT?

**Comment by takezoe** [comment]
@cvogt Great idea, I'm happy if I can maintain my code as the sub-project of Slick.

> we can also make any changes necessary to Slick itself in order to allow the api to be as source compatible to slick 2 as possible - e.g. allow bringing back "withSession". WDYT?

It would help existing Slick2 user who want to move to Slick3 like us. I would like to propose some API changes to improve Slick2 compatibility in the blocking API.

**Comment by cvogt** [comment]
Ok, I'll check with Typesafe to see if we can create a repo and get back to you. May take a week or so.

**Comment by szeiger** [comment]
I don't like the idea of bringing back the Slick 2 API. Getting rid of the old session handling makes things much simpler internally. And there's room for more simplification to be done. We should instead resurrect the blocking API from 3.0.0-M1, which was designed to work with the new API without duplication.

**Comment by takezoe** [comment]
> We should instead resurrect the blocking API from 3.0.0-M1, which was designed to work with the new API without duplication.

My implementation is just idea. I think resurrecting compatible API is better way than my ad-hoc implementation. Can I work for that?

**Comment by takezoe** [comment]
@szeiger "The blocking API from 3.0.0-M1" you mentioned is [this](https://github.com/slick/slick/commit/0ba0dd5efcdeaed38e9387e8fe4700f4a6b4b18e)? Hmm... It would not help us.

We love Slick's powerful type-safe query builder but don't need DBIO sometimes. DBIO is too difficult for non-functional programmers and async execution makes overhead in the small application. Furthermore, Slick2 has not been maintained already. These things are enough reason we avoid Slick.

We need blocking API without DBIO and executed on the same thread. I think the best way is we can choose DBIO or Slick2-like blocking API depending on the application. If Slick can't provide such API, we would like to publish that as a third-party library.

**Comment by deusaquilus** [comment]
IMHO, could this be a documentation issue? A section entitled 'Comparing Sync and Async Approaches' in dbio.html which basically recaps @szeiger's presentation (especially the slides at 23:50 and 33:00 [here](https://www.youtube.com/watch?v=WvxXz7aklik)), where you can see the blocking sync approach and the async for-comprehension approach side by side, would be very enlightening. There's a really good a-ha moment where @szeiger  says: 

> Hey, this is really similar, especially in scala!

In fact, I had absolutely no clue how the for-comprehensions were supposed to be used in the async API until I saw an example of the testing [here](https://github.com/lisinge/slick/blob/master/slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/JoinTest.scala). If you guys want new developers to switch to these systems/frameworks you have to educate and enlighten them in how to use basic FP, otherwise they'll ask for ways to move back to imperative and synchronous approaches as is the case here.

**Comment by nafg** [comment]
FWIW I've heard from a number of people that Slick's documentation is
inadequate, especially for people new to Scala.

**Comment by nafg** [comment]
On Tue, Aug 9, 2016, 12:45 PM Naoki Takezoe <notifications@github.com>
wrote:

> @szeiger <https://github.com/szeiger> "The blocking API from 3.0.0-M1"
> you mentioned is this
> <https://github.com/slick/slick/commit/0ba0dd5efcdeaed38e9387e8fe4700f4a6b4b18e>?
> Hmm... It would not help us.
>
> We love Slick's powerful type-safe query builder but don't need DBIO
> sometimes. DBIO is too difficult for non-functional programmers
>

Could that be solved with documentation?


and async execution makes overhead in the small application.
>

Perhaps that could be solved by changing a configuration?

Furthermore, Slick2 has not been maintained already.
>

What do you mean?

These things are enough reason we avoid Slick.

**Comment by deusaquilus** [comment]
Sadly DBIO is just the tip of the iceberg. When I first started using Slick there were dozens of brutal wtf moments. What the heck is a projection? Why are they useful/necessary? What the heck is the <> operator and why is it needed? What is an option-mapper and why do I implicitly need to pass it everywhere? What the heck is a 'lifted' type and why are they so important? How about some examples of these 'projections' and how they can be used beyond just mapping things to tuples/hlists? How about an in-depth architecture overview of this framework that doesn't presuppose you can trivially understand monads? Sorry if I'm derailing this conversation but if you guys want more adoption from OO devs, there has to be significantly more educational material then what currently exists.

**Comment by takezoe** [comment]
@nafg 

> and async execution makes overhead in the small application.
> Perhaps that could be solved by changing a configuration?

It might be true. This problem might be solved by providing an execution context that runs DBIO on the same thread.

> Furthermore, Slick2 has not been maintained already.
>What do you mean?

Sorry, my words was not enough. If Slick2 is active, we can choose use Slick2 instead of Slick3 as a blocking database access library.

Slick generates complex and slow SQL like [this](https://github.com/gitbucket/gitbucket/issues/1244#issuecomment-237449586). Of course, I know the query compiler is improving in Slick3, but they are not backported to Slick2. Therefore we can't choose Slick2 at the moment.

In addition, we have some Slick2 based applications. We have to decide whether we move existing Slick2 applications to Slick3 or other library.

**Comment by nafg** [comment]
On Tue, Aug 9, 2016 at 8:25 PM Naoki Takezoe <notifications@github.com>
wrote:

> @nafg <https://github.com/nafg>
>
> and async execution makes overhead in the small application.
> Perhaps that could be solved by changing a configuration?
>
> It might be true. This problem might be solved by providing an execution
> context that runs DBIO on the same thread.
>
> Furthermore, Slick2 has not been maintained already.
> What do you mean?
>
> Sorry, my words was not enough. If Slick2 is active, we can choose use
> Slick2 instead of Slick3 as a blocking database access library.
>
> Slick generates complex and slow SQL like this
> <https://github.com/gitbucket/gitbucket/issues/1244#issuecomment-237449586>.
> Of course, I know the query compiler is improving in Slick3, but they are
> not backported to Slick2. Therefore we can't choose Slick2 at the moment.
>
> In addition, we have some Slick2 based applications. We have to decide
> whether we move existing Slick2 applications to Slick3 or other libraryyou
>

One advantage of moving to Slick 3 is you could do it gradually: Update to
slick 3.0.x, which still has the session API but it's deprecated, and then
only change bits over gradually. Of course you can't get to 3.1.x (or
greater) gradually, but once you are fully on non-deprecated API you can
just upgrade further.

I think there it comes down to broader problems. For instance, having a
dedicated project leader and having a community are both needed. It seems
like for most of Slick's lifetime it had good leadership but weak
community. It seems like only after Lightbend needed to put Stefan on
scalac, they were successful at creating a team of community collaborators.
It would be nice to have both somehow. :) Somehow all of this feedback was
not really being given earlier on, or else it wasn't really heard?
Also a lot of people invested in learning Slick because of it being an
official Typesafe project, but now there doesn't seem to be anyone really
being paid to really move things forwards.
But like other Lightbend -- and many more -- products, maybe what it needs
most is documentation. If you have excellent documentation you will get a
lot more buyers, who will then become community and drive the project
forwards. Without good enough documentation a lot of potential community
will just go somewhere else.

**Comment by nafg** [comment]
Btw @takezoe perhaps this https://github.com/pelotom/effectful can make it easier to use the monadic API

**Comment by deusaquilus** [comment]
I agree with you @nafg but I would rephrase "they just go somewhere else" to "they will just keep using what they've been using before." Keep in mind that there are a lot of old players in the RBDMS space and moving developers away from the likes of Hibernate and plain SQL is akin to displacing a whale: doable, but tough. Slick is unique in it's ability to do object-relational mapping without the woes of N+1 as well as being able to polymorphically build queries of extraordinary expressiveness that could only otherwise be done in plain SQL. The former means that Slick/Play can compete in the Spring/EJB space which is obvious, the latter however means that Slick can replace much of the plain-sql based corporate reporting/business-intelligence infrastructure that feeds systems like Pentaho (this is a huge headache for corporations to manage and a giant potential market for Slick!). There is no other existing technology today that can effectively do both of these things, this is Slick's undeniable strength. Lack of good examples and a steep learning curve are unfortunately it's greatest weakness.

**Comment by takezoe** [comment]
I agree with @nafg and @deusaquilus in many points.

However this issue is about bringing back Slick2's blocking API. If Slick community does not want that, we should close this issue and raise an other issue.

**Comment by takezoe** [comment]
Anyway, I started a project to provide blocking API for Slick 3.1 as my personal project:
https://github.com/takezoe/blocking-slick

Please check it if you are interested in :-)

**Comment by takezoe** [comment]
Most of Slick2's blocking APIs are supported in the latest version of blocking-slick. There are some dirty implementation is remained because blocking-slick is implemented as an external library completely.

In addition, this is result of moving GitBucket which is existing Slick2 based application to Slick3 using blocking-slick. It was possible to move to Slick3 with small modification:
https://github.com/gitbucket/gitbucket/pull/1269/files

I will continue to maintain blocking-slick library above for moving our applications to Slick3.

**Comment by takezoe** [comment]
@cvogt @szeiger You should close this issue if there is no plan to provide slick 2's blocking API. Our blocking-slick library which was mentioned above is now available and stable enough.

**Comment by szeiger** [comment]
Agreed. I still think we should add the blocking methods for DBIO but this ticket is about the Slick 2 API. Thanks, @takezoe!

**Comment by cvogt** [comment]
added @takezoe's lib to http://slick.lightbend.com/docs/third-party-slick-3

**Comment by takezoe** [comment]
@cvogt Thanks!
